src/server/transaction.cc

```diff
@@ -847,90 +875,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
     DVLOG(1) << "Short-circuit ExecuteAsync " << DebugId();
     EngineShard::tlocal()->PollExecution("exec_cb", this);
-    intrusive_ptr_release(this);  // against use_count_.fetch_add above.
+    run_barrier_.Wait();  // End phase
     return;
   }
 
-  // We verify seq lock has the same generation number. See below for more info.
-  auto cb = [seq, this] {
-    EngineShard* shard = EngineShard::tlocal();
+  // Keep alive until all poll jobs finish. Function might exit earlier as we can be picked up
+  // by a foreign PollExecution.
+  use_count_.fetch_add(unique_shard_cnt_, memory_order_relaxed);
+
+  auto poll_cb = [phase, this] {
+    // As we might have been handled by a foregin poll job, we first need to verify that:
+    // 1. The phase is still active
+    // 2. We're still armed (i.e. nobody else run us on this shard)
+    // 3. We're still in the same phase as we expect (i.e. we didn't miss a full cycle)
 
-    bool is_armed = IsArmedInShard(shard->shard_id());
-    // First we check that this shard should run a callback by checking IsArmedInShard.
-    if (is_armed) {
-      uint32_t seq_after = seqlock_.load(memory_order_relaxed);
-
-      DVLOG(3) << "PollExecCb " << DebugId() << " sid(" << shard->shard_id() << ") "
-               << run_count_.load(memory_order_relaxed);
-
-      // We also make sure that for multi-operation transactions like Multi/Eval
-      // this callback runs on a correct operation. We want to avoid a situation
-      // where the first operation is executed and the second operation is armed and
-      // now this callback from the previous operation finally runs and calls PollExecution.
-      // It is usually ok, but for single shard operations we abuse index 0 in shard_data_
-      // Therefore we may end up with a situation where this old callback runs on shard 7,
-      // accessing shard_data_[0] that now represents shard 5 for the next operation.
-      // seqlock provides protection for that so each cb will only run on the operation it has
-      // been tasked with.
-      // We also must first check is_armed and only then seqlock. The first check ensures that
-      // the coordinator thread crossed
-      // "run_count_.store(unique_shard_cnt_, memory_order_release);" barrier and our seqlock_
-      // is valid.
-      if (seq_after == seq) {
-        // shard->PollExecution(this) does not necessarily execute this transaction.
-        // Therefore, everything that should be handled during the callback execution
-        // should go into RunInShard.
+    EngineShard* shard = EngineShard::tlocal();
+    if (IsArmedInShard(shard->shard_id())) {
+      if (run_barrier_.IsMyPhase(phase)) {
+        // PollExecution doesn't necessarily this transaction, so execution and cleanup code belongs
+        // to RunInShard
         shard->PollExecution("exec_cb", this);
       } else {
         VLOG(1) << "Skipping PollExecution " << DebugId() << " sid(" << shard->shard_id() << ")";
       }
     }
```

Comment:
My radical opinion: is_armed doesn't have to be a separate atomic and can be a regular shard flag.

How do we read safely from it without exiting the execution phase at the same time?

1. Atomically increment `use_count` and enter the section if it's larger than 1
2. Because we added an artifical job, Wait() is guaranteed to be waiting for us
3. Safely read is_armed and determine here whether we need to run -> profit

Same logic applies to the phase, it's changed only after the execution phase, so it's safe to read while inside it